content/blog/2022-05-08-Apache-ShenYu(incubating)-Integrated-Skywalking-Practice-Observability/index.md

```diff
@@ -0,0 +1,130 @@
+---
+
+### Content
+
+1. [Introduction of SkyWalking and ShenYu](#1.-Introduction-of-SkyWalking-and-ShenYu)
+2. [Apache ShenYu plugin implementation principle](#2.-Apache-ShenYu-plugin-implementation-principle)
+3. [Adding generalized call tracking to the gRPC plugin and keeping it compatible](#3.-Adding-generalized-call-tracking-to-the-gRPC-plugin-and-keeping-it-compatible)
+4. [ShenYu Gateway Observability Practice](#4.-ShenYu-Gateway-Observability-Practice)
+5. [Summary](#5.-Summary)
+
+
+## 1. Introduction of SkyWalking and ShenYu
+
+### 1.1 SkyWalking
+
+[SkyWalking](https://github.com/hutaishi/skywalking) is an Application Performance Monitoring (APM) and Observability Analysis Platform (OAP) for microservices, distributed systems, and cloud natives,
+Has powerful features that provide a multi-dimensional means of application performance analysis, including distributed topology diagrams, application performance metrics, distributed link tracing, log correlation analysis and alerts. Also has a very rich ecology. Widely used in various companies and open source projects.
+
+### 1.2 Apache ShenYu (incubating)
+
+[Apache ShenYu (incubating)](https://github.com/apache/incubator-shenyu)
+High-performance,multi-protocol,extensible,responsive API Gateway. Compatible with a variety of mainstream framework systems, support hot plug, 
+users can customize the development, meet the current situation and future needs of users in a variety of scenarios, experienced the temper of large-scale scenes.
+Rich protocol support: `Http`, `Spring Cloud`, `gRPC`, `Dubbo`, `SOFARPC`, `Motan`, `Tars`, etc.
+
+![](shenyu-arch.png)
+
+
+## 2. Apache ShenYu plugin implementation principle
+
+ShenYu's asynchrony is a little different from previous exposure to asynchrony, it is a full-link asynchrony, the execution of each plug-in is asynchronous, and thread switching is not a single fixed situation (and the individual plug-in implementation is related).
+The gateway initiates service calls of various protocol types, and the existing SkyWalking plugins create ExitSpan (synchronous or asynchronous) when they initiate service calls.  The gateway receives the request and creates an asynchronous EntrySpan.
+The asynchronous EntrySpan needs to be concatenated with the synchronous or asynchronous ExitSpan, otherwise the link will be broken.
+
+
+There are 2 types of tandem solutions：
+- **Snapshot Delivery**： Pass the snapshot after creating the EntrySpan to the thread that created the ExitSpan in some way.  
+  Currently this approach is used in the asynchronous WebClient plugin, which can receive asynchronous snapshots. shenYu proxy Http service or SpringCloud service is to achieve span concatenation through snapshot passing.
+- **LocalSpan transit.**
+  Other Rpc class plugins do not receive snapshots for concatenation like Asynchronous WebClient. Although you can modify other Rpc plugins to receive snapshots for concatenation, it is not recommended or necessary to do so.
```

Comment:
Rpc should be RPC.

Reply:
> Rpc should be RPC.

I did not pay attention to this detail, has been modified